File: 0740-delete-and-earn/0740-delete-and-earn.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        int max = 0;
        for (int num : nums) {
            if (num > max) max = num;
        }
        int[] points = new int[max + 1];
        for (int num : nums) {
            points[num] += num;
        }
        
        int prev = 0;
        int curr = 0;
        for (int point : points) {
            int next = Math.max(curr, prev + point);
            prev = curr;
            curr = next;
        }
        return curr;
    }
}

File: 0740-delete-and-earn/0740-delete-and-earn.py
class Solution(object):
    def deleteAndEarn(self, nums):
        max_num = max(nums)
        freq = [0] * (max_num + 1)

        for num in nums:
            freq[num] += num
        
        take, notTake = 0, 0
        for fr in freq:
            tmp = take
            take = notTake + fr
            notTake = max(tmp, notTake)
        
        return max(take, notTake)
